### PR TITLE
Allow name of blackbox resource .f file to change from static value

### PR DESCRIPTION
**Pull request by albert-magyar**
Currently, the `.f` resource file gets generated with the constant filename of `firrtl_black_box_resource_files.f`. This PR keeps that as a default, but allows the user to override this filename with an optional annotation. The only other visibility outside of the transform is that a new parameter is added to `verilogToCpp` to enable sourcing a renamed file (but the default filename is used by default).

This is helpful to manage multiple `.f` files, which has come up as an issue in Chipyard.

### CODE REVIEW

**Comment by albert-magyar** [comment]
@colinschmidt I'm going to try using this in Chipyard before I take the DO NOT MERGE off

File: src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala [hunk 1]
```diff
@@ -150,17 +159,18 @@ object BlackBoxSourceHelper {
     out.close()
   }
 
-  val fileListName = "firrtl_black_box_resource_files.f"
```
Comment (jackkoenig) [inline src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala]: I think it'd be good to have `fileListName` as a deprecated `def` just in case someone is using it

Comment (albert-magyar) [reply src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala]: Good call

**Comment by jerryz123** [comment]
Resolves #1090

**Comment by albert-magyar** [comment]
@colinschmidt, this works to resolve the current `.f` issues in chipyard.

The one question left is whether to make the string filename in the annotation a) absolute or b) relative to `targetDir`. It seems like more existing FIRRTL things treat overridden filenames as (a) than (b), but maybe @jackkoenig or @azidar could chime in.

**Comment by seldridge** [comment]
Have you tried using `firrtl.options.StageOptions.getBuildFileName`? That will take a string and do absolute if it's absolute (in Linux if it starts with `/`) and relative to the target directory otherwise. 

Something like the following would likely work (untested, though!):

```scala
import firrtl.AnnotationSeq
import firrtl.options.{StageOptions, Viewer}

abstract class Foo {
  /** Given some annotations */
  val annotations: AnnotationSeq

  /** Construct [[StageOptions]] from the annotations */
  val sopts = Viewer[StageOptions].view(annotations)

  /** An absolute path */
  sopts.getBuildFileName("foo/bar/baz")

  /** A path relative to the target directory as determined by the annotations */
  sopts.getBuildFileName("/foor/bar/baz")
}
```

**Comment by jackkoenig** [comment]
I think @seldridge's suggestion is good; supporting both use styles of paths is the most flexible approach.

I do want to make a point about avoiding absolute paths to whatever extent possible. Historically we (myself especially) have done a poor job of this, because absolute paths seem better at first blush, but they're actually pretty problematic. In particular, it is incredibly difficult to make a build relocatable and reproducible if you use absolute paths. 

### Relocatability

Imagine if your annotations are full of absolute file paths, but then you want to farm out part of your compute to some other machine in the cloud. If the machine's directory structure doesn't match the directory structure you used to create the paths in the annotations, you're hosed. 

### Caching

Many tools now use hash based caching (including sharing caching between users) that can enable powerful short-circuiting builds when some intermediate artifact is actually the same as a previous run (see bazel, pants, wake). Absolute paths make this nigh on impossible because they differ between users or even for a single user when you have multiple clones.

**TLDR** We should use relative paths.

**Comment by jwright6323** [comment]
@seldridge are these backwards?

>   /** An absolute path */
>   val absolutePath: String = sopts.getBuildFileName("foo/bar/baz")
> 
>   /** A path relative to the target directory as determined by the annotations */
>   val relativePath: String = sopts.getBuildFileName("/foor/bar/baz")

**Comment by seldridge** [comment]
@jwright6323: Yep, backwards.

**Comment by albert-magyar** [comment]
@seldridge that is a good idea. Unfortunately, `BlackBoxSourceHelper` has it's _own distinct_ `targetDir` option (so blackbox resources could end up in a different place), so I basically copied the functionality of `getBuildFileName`.

This now supports absolute and relative paths.

**Comment by seldridge** [comment]
That sounds fine. At some point I'd like to revisit and get things unified under `TargetDirAnnotation`. Can happen in the future...
